### PR TITLE
Move log_normalize from hdpmodel to matutils, add support for feature normalization

### PR DESCRIPTION
**Pull request by ArifAhmed1995**
- Log normalizing may be used for other models too, hence it's better to shift it to matutils
- log normalize : Currently for an input 2D array , independently normalizing each sample is done.An optional axis parameter is introduced to give support for normalizing along feature.

See axis parameter in : 
http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html

### CODE REVIEW

**Comment by tmylk** [comment]
dw#

**Comment by tmylk** [comment]
Thanks for the PR!

File: gensim/matutils.py [hunk 1]
```diff
@@ -326,6 +326,29 @@ def ret_normalized_vec(vec, length):
...
+def ret_log_normalize_vec(vec, axis = 1):
+    log_max = 100.0
+    if len(vec.shape) == 1:
+        max_val = numpy.max(vec)
+        log_shift = log_max - numpy.log(len(vec) + 1.0) - max_val
+        tot = numpy.sum(numpy.exp(vec + log_shift))
+        log_norm = numpy.log(tot) - log_shift
+        vec = vec - log_norm
+    else:
+        if axis == 1:#independently normalize each sample
```
Comment (piskvorky) [inline gensim/matutils.py]: PEP8: at least two spaces between code and inline comment.

File: gensim/matutils.py [hunk 2]
```diff
@@ -326,6 +326,29 @@ def ret_normalized_vec(vec, length):
...
+            max_val = numpy.max(vec, 1)
+            log_shift = log_max - numpy.log(vec.shape[1] + 1.0) - max_val
+            tot = numpy.sum(numpy.exp(vec + log_shift[:, numpy.newaxis]), 1)
+            log_norm = numpy.log(tot) - log_shift
+            vec = vec - log_norm[:, numpy.newaxis]
+        elif axis == 0:#normalize each feature
```
Comment (piskvorky) [inline gensim/matutils.py]: dtto

File: gensim/matutils.py [hunk 3]
```diff
@@ -326,6 +326,29 @@ def ret_normalized_vec(vec, length):
     else:
         return list(vec)
 
+def ret_log_normalize_vec(vec, axis = 1):
```
Comment (piskvorky) [inline gensim/matutils.py]: PEP8: no space around function parameter values.

File: gensim/matutils.py [hunk 4]
```diff
@@ -326,6 +326,29 @@ def ret_normalized_vec(vec, length):
...
+            max_val = numpy.max(vec, 1)
+            log_shift = log_max - numpy.log(vec.shape[1] + 1.0) - max_val
+            tot = numpy.sum(numpy.exp(vec + log_shift[:, numpy.newaxis]), 1)
+            log_norm = numpy.log(tot) - log_shift
+            vec = vec - log_norm[:, numpy.newaxis]
+        elif axis == 0:#normalize each feature
+            k = ret_log_normalize_vec(vec.T)
+            return (k[0].T, k[1])
+        else:
+            raise ValueError("'%d' is not a supported axis" % axis)
```
Comment (piskvorky) [inline gensim/matutils.py]: Better format as `%s`. In case user inadvertently passes a parameter of different type, they should still get a sensible exception.
